neps/nep-0371.md

```diff
@@ -0,0 +1,149 @@
+---
+NEP: 371
+Title: Runtime Function Call Access Key Info
+Author: Ben Kurrek <ben.kurrek@near.org>, Jakob Meier <jakob@near.org>
+DiscussionsTo: https://gov.near.org/t/expose-gas-price-and-contract-access-key-info-within-runtime/24788
+Status: Draft
+Type: Standards Track
+Category: Review
+Created: 13-Jul-2022
+---
+
+## Summary
+
+There is currently no way to query for information about access keys stored on a contract at runtime. Adding the host function `access_key_info` to the runtime specification will allow to query this for the current account an easy way. This fully covers the most important use cases that are currently known for querying access keys at runtime.
+
+## Motivation
+
+As contracts start to use access keys in more creative ways, developers will want to know information such as the access key's allowance, receiver and more. Since contracts are often built for scalability, Gas and storage costs are often delegated to the end user. If you charge the user for an access key and its allowance, when the key is deleted or used up, that allowance should be refunded to the purchaser. Currently, there is no way to query for the leftover allowance before a key is deleted. This information can tell you how much Gas was spent, and allows for a slew of different scenarios to arise.
+
+## Rationale and alternatives
+
+Without creating a promise, we can only query for information stored on the same shard. This is why in this proposal, the access key information at runtime will only come from keys stored on the current contract. We had also investigated the possibility of introducing pagination for all keys stored on the contract but this is infeasible at the moment due to the significant redesign requirement of the VM logic.
+
+## Runtime specification
+
+The contract runtime provides one host function called `access_key_info` which is callable in regular calls and views.
+
+```rust
+fn access_key_info (
+    public_key_len: u64,                // in:  number of bytes of the public key to query
+    public_key_ptr: u64,                // in:  pointer to the public key to query
+    nonce_ptr: u64                      // out: pointer to where a u64 can be written by the host
+    allowance_ptr: u64,                 // out: pointer to where a u128 can be written by the host
+    account_id_register_id: u64,        // out: register to write the access key holder account id
+    method_names_register_id: u64,      // out: register to write the method names as a comma separated list
```

Comment:
Register API is specified here: https://nomicon.io/RuntimeSpec/Components/BindingsSpec/RegistersAPI

It looks like it is not documented / specified but the upper limit of a register is set by a runtime parameter, `max_register_size: 104_857_600`. Plus, the sum of all register sizes is limited by `registers_memory_limit: 1_073_741_824`. But both of these are currently only theoretical limits, as writing 1GiB into registers would cost around 2925 Tgas and writing 100MiB costs around 285 Tgas. (But I should still include it as a failure mode in the specs, thanks for pointing it out!)

The upper limit on the total length of all method names is limited by the runtime parameter `max_number_bytes_method_names: 2_000`, which is enforced when creating an access key. Those 2000 bytes can be split up into method names and separators arbitrarily. (Each valid method name is also limited to 256 bytes but that's not really relevant here, since no validity checks are made on the method names of an access key.)